public/app/features/scopes/selector/ScopesSelectorService.ts

```diff
@@ -0,0 +1,393 @@
+      const possibleParent = childNodesArr.find((childNode) => childNode.selectable && childNode.linkId === scopeName);
+
+      if (possibleParent) {
+        acc[scopeName] = [...path, possibleParent.name];
+      }
+
+      return acc;
+    }, {});
+
+    // Update the paths of the selected scopes based on what we found
+    const newSelectedScopes = selectedScopes.map((selectedScope) => {
+      if (selectedScope.path.length > 0) {
+        return selectedScope;
+      }
+
+      return {
+        ...selectedScope,
+        path: scopeNamesWithPaths[selectedScope.scope.metadata.name] ?? [],
+      };
+    });
+
+    // Update the paths of the tree scopes based on what we found
+    const newTreeScopes = treeScopes.map((treeScope) => {
+      if (treeScope.path.length > 0) {
+        return treeScope;
+      }
+
+      return {
+        ...treeScope,
+        path: scopeNamesWithPaths[treeScope.scopeName] ?? [],
+      };
+    });
+
+    return [newSelectedScopes, newTreeScopes];
+  };
+
+  public fetchNodeApi = async (parent: string, query: string): Promise<NodesMap> => {
+    try {
+      const nodes =
+        (
+          await getBackendSrv().get<{ items: ScopeNode[] }>(
+            `/apis/${this._apiGroup}/${this._apiVersion}/namespaces/${this._apiNamespace}/find/scope_node_children`,
+            { parent, query }
+          )
+        )?.items ?? [];
+
+      return nodes.reduce<NodesMap>((acc, { metadata: { name }, spec }) => {
+        acc[name] = {
+          name,
+          ...spec,
+          expandable: spec.nodeType === 'container',
+          selectable: spec.linkType === 'scope',
+          expanded: false,
+          query: '',
+          reason: NodeReason.Result,
+          nodes: {},
+        };
+        return acc;
+      }, {});
+    } catch (err) {
+      return {};
+    }
+  };
+
+  public fetchScopeApi = async (name: string): Promise<Scope> => {
+    if (this._scopesCache.has(name)) {
+      return this._scopesCache.get(name)!;
+    }
+
+    const response = new Promise<Scope>(async (resolve) => {
+      const basicScope = this.getBasicScope(name);
+
+      try {
+        const serverScope = await getBackendSrv().get<Scope>(
+          `/apis/${this._apiGroup}/${this._apiVersion}/namespaces/${this._apiNamespace}/scopes/${name}`
+        );
+
+        const scope = {
+          ...basicScope,
+          ...serverScope,
+          metadata: {
+            ...basicScope.metadata,
+            ...serverScope.metadata,
+          },
+          spec: {
+            ...basicScope.spec,
+            ...serverScope.spec,
+          },
+        };
+
+        resolve(scope);
+      } catch (err) {
+        this._scopesCache.delete(name);
+
+        resolve(basicScope);
+      }
+    });
+
+    this._scopesCache.set(name, response);
+
+    return response;
+  };
+
+  public fetchScopesApi = async (treeScopes: TreeScope[]): Promise<SelectedScope[]> => {
```

Comment:
Do these fetch api calls need to be public?

Reply:
It has be for the tests, unfortunately 😢 Since this is in core, I don't think it matters too much right now.

Reply:
yeah makes sense. For some reason I thought jest would continue spying on that method even if it's private since.